### DISCUSSION TITLE
How do you set the binary file icon?

### DISCUSSION THREAD

**Discussion by MrTanoshii** [emilk/egui#2026] [Q&A] [answered]
Hey guys, I managed to set the app icon (window decorator and taskbar) but I don't know how to set the binary file icon.
Does anyone have a solution please?

**Comment by nozwock** [comment, marked as answer]
This is not really an `egui` question but Icon resources for **windows binaries** can easily be set using the [winres](https://docs.rs/winres/latest/winres/) crate.

Though for Rust 1.61 and above, the crate [doesn't work](https://github.com/mxre/winres/issues/40), so you'll have to use a [fork](https://github.com/Nilstrieb/winres) containing the fix instead.
So just add the forked `winres` as a build dependency in your `Cargo.toml`:
```toml
[build-dependencies]
winres = { git = "https://github.com/Nilstrieb/winres", branch = "linking-flags" }
```

After that, you'll also need to add winres metadata section to your `Cargo.toml`, see [winres](https://github.com/mxre/winres#additional-options) for details.
The section can be left completely empty if you want though, I had to add it since `winres` fails for me if I don't have its section in the `Cargo.toml`.

Then, at last you'll need to add a [build script](https://doc.rust-lang.org/cargo/reference/build-scripts.html) (a file `build.rs` in the root of your crate):
```rust
use std::io;
use winres::WindowsResource;

// for eg. this is my build script
// I've tested this on linux and windows with rust 1.64

fn main() -> io::Result<()> {
    if std::env::var("CARGO_CFG_TARGET_FAMILY").unwrap() == "windows" {
        let mut res = WindowsResource::new();
        match std::env::var("CARGO_CFG_TARGET_ENV").unwrap().as_str() {
            "gnu" => {
                res.set_ar_path("x86_64-w64-mingw32-ar")
                    .set_windres_path("x86_64-w64-mingw32-windres");
            }
            "msvc" => {}
            _ => panic!("unsupported env"),
        };
        res.set_icon("assets/icon.ico");
        res.compile()?;
    }
    Ok(())
}
```

This [here](https://docs.rs/winres/latest/winres/#example) is the example build script on `winres` crate doc.

Take a look at these for more `build.rs` examples:
https://github.com/mxre/winres/issues/33
https://github.com/mxre/winres/pull/24

Edit: Fix typos

**Reply by MrTanoshii** [reply]
Thank you so much!
